lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -932,10 +792,23 @@ void LCIStrategicPlugin::planWhenAPPROACHING(const cav_srvs::PlanManeuversReques
 
   ROS_DEBUG_STREAM("SPEED PROFILE CASE:" << ts_params.case_num);
 
-  // CASE SELECTION END
+  /////////////  4. Safety Check against traffic signals and Final Maneuver Generation //////////////
+
+   double safe_distance_to_stop = pow(current_state.speed, 2)/(2 * max_comfort_decel_norm_);
+  ROS_DEBUG_STREAM("safe_distance_to_stop at max_comfort_decel:  " << safe_distance_to_stop << ", max_comfort_decel_norm_: " << max_comfort_decel_norm_);
+
+  double desired_distance_to_stop = pow(current_state.speed, 2)/(2 * max_comfort_decel_norm_ * config_.deceleration_fraction) + config_.desired_distance_to_stop_buffer;
+  ROS_DEBUG_STREAM("desired_distance_to_stop at: " << desired_distance_to_stop << ", where effective deceleration rate is: " << max_comfort_decel_norm_ * config_.deceleration_fraction);
+  
+  desired_distance_to_stop = std::max(desired_distance_to_stop, config_.stopping_location_buffer);
+
+  ROS_DEBUG_STREAM("new desired_distance_to_stop: " << desired_distance_to_stop);
+
+  ROS_DEBUG_STREAM("distance_remaining_to_traffic_light:  " << distance_remaining_to_traffic_light << ", current_state.speed: " << current_state.speed);
 
   // Although algorithm determines nearest_green_time is possible, check if the vehicle can arrive with certainty (Case 1-7)
-  if (ts_params.is_algorithm_successful && ts_params.case_num != TSCase::CASE_8) 
+  if (ts_params.is_algorithm_successful && ts_params.case_num != TSCase::CASE_8 && 
+    (distance_remaining_to_traffic_light >= desired_distance_to_stop || !in_tdb)) // Given ET is in TDB, but vehicle is too close to intersection
```

Comment:
We should separate ET_in_TBD vs ET_in_GREEN in this case because the platform shouldn't generate handleGreenManeuver when distance is too close. While in ET_in_GREEN however, it should always try to generate it, so the 2 flags should not be lumped together.

Reply:
What I think is that by adding "is_entry_time_within_green_or_tdb == true", we can be sure that ET is not within red or yellow. It confirms that ET is within green or TBD and the "in_tbd" parameter confirms if ET is in TBD or not. Otherwise, if "in_tbd" is false, ET could be within red and we still run the logic in this if condition, right?

Reply:
Ah that is true. The old logic would have checked certainty case and would have detected the RED, but I went ahead and added the check to make it more clear to readers